src/pyflame.cc

```diff
@@ -108,63 +167,43 @@ int main(int argc, char **argv) {
-      std::unordered_map<frames_t, size_t, FrameHash> buckets;
-      size_t idle = 0;
-      auto end =
-          std::chrono::system_clock::now() +
-          std::chrono::microseconds(static_cast<long>(seconds * 1000000));
-      for (;;) {
-        const unsigned long frame_addr = FirstFrameAddr(pid, tstate_addr);
-        if (frame_addr == 0) {
-          if (include_idle) {
-            idle++;
-          }
-        } else {
-          frames_t frames = GetStack(pid, frame_addr);
-          auto it = buckets.find(frames);
-          if (it == buckets.end()) {
-            buckets.insert(it, {frames, 1});
-          } else {
-            it->second++;
+    const std::chrono::microseconds interval{
+        static_cast<long>(sample_rate * 1000000)};
+    std::vector<FrameTS> call_stacks;
+    size_t idle = 0;
+    auto end = std::chrono::system_clock::now() +
+               std::chrono::microseconds(static_cast<long>(seconds * 1000000));
+    for (;;) {
+      const unsigned long frame_addr = FirstFrameAddr(pid, tstate_addr);
+      if (frame_addr == 0) {
+        if (include_idle) {
+          idle++;
+          /* Time stamp empty call stacks only if required. Since lots of time
+           * the process will be idle, this is a good optimization to have */
+          if (include_ts) {
+            auto now = std::chrono::system_clock::now();
+            call_stacks.push_back({now, {}});
           }
         }
+      } else {
+        frames_t frames = GetStack(pid, frame_addr);
         auto now = std::chrono::system_clock::now();
-        if (now + interval >= end) {
-          break;
-        }
-        PtraceDetach(pid);
-        std::this_thread::sleep_for(interval);
-        PtraceAttach(pid);
+        call_stacks.push_back({now, frames});
       }
-      if (idle) {
-        std::cout << "(idle) " << idle << "\n";
+      auto now = std::chrono::system_clock::now();
+      if (now + interval >= end) {
+        break;
       }
-      // process the frames
-      for (const auto &kv : buckets) {
-        if (kv.first.empty()) {
-          std::cerr << "uh oh\n";
-          return 1;
-        }
-        auto last = kv.first.rend();
-        last--;
-        for (auto it = kv.first.rbegin(); it != last; ++it) {
-          std::cout << *it << ";";
-        }
-        std::cout << *last << " " << kv.second << "\n";
+      PtraceDetach(pid);
+      std::this_thread::sleep_for(interval);
+      PtraceAttach(pid);
+    }
+    if (!include_ts) {
+      if (0 != PrintFrames(call_stacks, idle)) {
+        return 1;
       }
     } else {
-      const unsigned long frame_addr = FirstFrameAddr(pid, tstate_addr);
-      if (frame_addr) {
-        std::vector<Frame> stack = GetStack(pid, frame_addr);
-        for (auto it = stack.rbegin(); it != stack.rend(); it++) {
-          std::cout << *it << "\n";
-        }
-      } else {
-        std::cout << "(idle)\n";
-      }
+      (void)PrintFramesTS(call_stacks);
```

Comment:
(void) is an explicit way of saying i am not deliberately checking the return value, also to keep linters happy if we use them